File: dev/20190822/k2works/src/Domain/Type/FizzBuzzType.ts
import { IFizzBuzzType } from "../../Application/Service/IFizzBuzzType";
import { FizzBuzzType01 } from "./FizzBuzzType01";
import { FizzBuzzType02 } from "./FizzBuzzType02";
import { FizzBuzzType03 } from "./FizzBuzzType03";

export class FizzBuzzType {
  public static readonly one = new FizzBuzzType01();
  public static readonly two = new FizzBuzzType02();
  public static readonly three = new FizzBuzzType03();
  public static valueOf(type: string): IFizzBuzzType {
    switch (type) {
      case "one":
        return FizzBuzzType.one;
      case "two":
        return FizzBuzzType.two;
      case "three":
        return FizzBuzzType.three;
      default:
        return FizzBuzzType.one;
    }
  }
}

File: dev/20190902/k2works/app/FizzBuzz/Domain/Type/FizzBuzzType.cs
using System;
using FizzBuzz.Application.Command;

namespace FizzBuzz.Domain.Type
{
    public abstract class FizzBuzzType: Enumeration
    {
        public static FizzBuzzType One = new FizzBuzzType01();
        public static FizzBuzzType Two = new FizzBuzzType02();
        public static FizzBuzzType Three = new FizzBuzzType03();
        
        public static readonly string FIZZ = "Fizz";
        public static readonly string BUZZ = "Buzz";
        public static readonly string FIZZ_BUZZ = "FizzBuzz";
        public abstract IFizzBuzz Generate(int number);
    }
    public abstract class Enumeration : IComparable
    {
        public int CompareTo(object obj)
        {
            throw new NotImplementedException();
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(this, obj))
            {
                return true;
            }

            if (ReferenceEquals(obj, null))
            {
                return false;
            }

            throw new NotImplementedException();
        }

        public override int GetHashCode() => throw new NotImplementedException();

        public static bool operator ==(Enumeration left, Enumeration right)
        {
            if (ReferenceEquals(left, null))
            {
                return ReferenceEquals(right, null);
            }

            return left.Equals(right);
        }

        public static bool operator !=(Enumeration left, Enumeration right)
        {
            return !(left == right);
        }

        public static bool operator <(Enumeration left, Enumeration right)
        {
            return ReferenceEquals(left, null) ? !ReferenceEquals(right, null) : left.CompareTo(right) < 0;
        }

        public static bool operator <=(Enumeration left, Enumeration right)
        {
            return ReferenceEquals(left, null) || left.CompareTo(right) <= 0;
        }

        public static bool operator >(Enumeration left, Enumeration right)
        {
            return !ReferenceEquals(left, null) && left.CompareTo(right) > 0;
        }

        public static bool operator >=(Enumeration left, Enumeration right)
        {
            return ReferenceEquals(left, null) ? ReferenceEquals(right, null) : left.CompareTo(right) >= 0;
        }
    }
}